### ISSUE TITLE
High cpu load and performance problem

### ISSUE THREAD

**Issue by pascalverlinden** [hyperledger/burrow#713]
**Version info**
Monax CLI Version: 0.17.0 & 0.18.0
Monax db 0.16.3
docker  17.03.0-ce 
OS: Ubuntu 14.04.5 LTS

**Description**
Creating many contracts leads to high CPU load and performance problems with the chain. After stopping the chain the CPU usage drops, but then it is no longer possible to start the chain as this leads to an error "panic: Paniced on a Sanity Check: Expected storeHeight (1867) and stateHeight (1866) to match", which seems to be a problem with Tendermint ( see https://github.com/tendermint/tendermint/issues/388 ). There has been a fix in Tendermint (version 0.9) but this version is not used yet by monax, apparently.

There might be a relation with the following issue: https://github.com/hyperledger/burrow/issues/611

**Steps to reproduce**
Example code is available in https://github.com/pascalverlinden/perfo

**Comment by joeriexelmans** [comment]
Updated example code to work with latest version of burrow and bosmarmot tooling:
https://github.com/pascalverlinden/perfo/tree/latest-version

Performance issues are still present:
- After creating 80 contracts of type 'D', and then doing nothing, CPU usage of burrow process remains at 20%. After there are 160 contracts on the chain, CPU usage is 40%.
- Latency of creating a new contract goes from 1s to 2s after about 60 contracts.
- CPU usage is close to 100% while sending a single transaction per block.

**Comment by silasdavis** [comment]
## Profiling
Here is a 30 second CPU profile from a fresh train humming along at < 1% CPU on my machine:

```
(pprof) top 40
Showing nodes accounting for 730ms, 96.05% of 760ms total
Showing top 40 nodes out of 233
      flat  flat%   sum%        cum   cum%
      70ms  9.21%  9.21%       70ms  9.21%  runtime.futex
      70ms  9.21% 18.42%       70ms  9.21%  runtime.usleep
      60ms  7.89% 26.32%       80ms 10.53%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.FeMul
      60ms  7.89% 34.21%       60ms  7.89%  github.com/hyperledger/burrow/vendor/golang.org/x/crypto/ripemd160._Block
      50ms  6.58% 40.79%       50ms  6.58%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.feSquare
      30ms  3.95% 44.74%       30ms  3.95%  runtime.findnull
      20ms  2.63% 47.37%       20ms  2.63%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.FeCombine
      20ms  2.63% 50.00%       20ms  2.63%  github.com/hyperledger/burrow/vendor/github.com/tendermint/go-wire.WriteTo
      20ms  2.63% 52.63%       60ms  7.89%  runtime.(*pcExpander).init
      20ms  2.63% 55.26%       20ms  2.63%  runtime.memmove
      20ms  2.63% 57.89%       20ms  2.63%  runtime.step
      10ms  1.32% 59.21%       10ms  1.32%  encoding/hex.Encode
      10ms  1.32% 60.53%       30ms  3.95%  fmt.(*pp).handleMethods
      10ms  1.32% 61.84%       10ms  1.32%  github.com/hyperledger/burrow/logging/loggers.(*ChannelLogger).WaitReadLogLine
      10ms  1.32% 63.16%       30ms  3.95%  github.com/hyperledger/burrow/logging/loggers.(*burrowFormatLogger).Log.func1
      10ms  1.32% 64.47%       20ms  2.63%  github.com/hyperledger/burrow/logging/loggers.(*infoTraceLogger).Info
      10ms  1.32% 65.79%       30ms  3.95%  github.com/hyperledger/burrow/logging/structure.Vectorise
      10ms  1.32% 67.11%       40ms  5.26%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.(*CompletedGroupElement).ToProjective
      10ms  1.32% 68.42%       40ms  5.26%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.(*ProjectiveGroupElement).Double
      10ms  1.32% 69.74%       10ms  1.32%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.FeAdd
      10ms  1.32% 71.05%       10ms  1.32%  github.com/hyperledger/burrow/vendor/github.com/tendermint/ed25519/edwards25519.FeSquare2
      10ms  1.32% 72.37%       90ms 11.84%  github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/types.(*Header).Hash
      10ms  1.32% 73.68%       70ms  9.21%  github.com/hyperledger/burrow/vendor/github.com/tendermint/tmlibs/merkle.KVPair.Hash
      10ms  1.32% 75.00%       10ms  1.32%  reflect.Value.Uint
      10ms  1.32% 76.32%       10ms  1.32%  runtime.(*fixalloc).alloc
      10ms  1.32% 77.63%       10ms  1.32%  runtime.(*mspan).refillAllocCache
      10ms  1.32% 78.95%       10ms  1.32%  runtime.(*waitq).dequeue
      10ms  1.32% 80.26%       10ms  1.32%  runtime.aeshashbody
      10ms  1.32% 81.58%       20ms  2.63%  runtime.assertE2I2
      10ms  1.32% 82.89%       10ms  1.32%  runtime.execute
      10ms  1.32% 84.21%       70ms  9.21%  runtime.findrunnable
      10ms  1.32% 85.53%       60ms  7.89%  runtime.futexsleep
      10ms  1.32% 86.84%       10ms  1.32%  runtime.getitab
      10ms  1.32% 88.16%       10ms  1.32%  runtime.makemap
      10ms  1.32% 89.47%       40ms  5.26%  runtime.mallocgc
      10ms  1.32% 90.79%       10ms  1.32%  runtime.mapaccess2_faststr
      10ms  1.32% 92.11%       10ms  1.32%  runtime.memclrNoHeapPointers
      10ms  1.32% 93.42%       10ms  1.32%  strings.indexFunc
      10ms  1.32% 94.74%       10ms  1.32%  sync.(*Once).Do
      10ms  1.32% 96.05%       10ms  1.32%  sync.(*Pool).pin
```


Here is a profile after running many_contracts to ~100 contracts when the chain is not receiving TXs and should be in steady state but has CPU pegged to ~40%:

```
(pprof) top40
Showing nodes accounting for 9.33s, 79.61% of 11.72s total
Dropped 306 nodes (cum <= 0.06s)
Showing top 40 nodes out of 189
      flat  flat%   sum%        cum   cum%
     1.94s 16.55% 16.55%      1.94s 16.55%  github.com/hyperledger/burrow/vendor/golang.org/x/crypto/ripemd160._Block
     0.99s  8.45% 25.00%      0.99s  8.45%  runtime.cmpbody
     0.71s  6.06% 31.06%      1.12s  9.56%  runtime.mallocgc
     0.61s  5.20% 36.26%      0.63s  5.38%  syscall.Syscall
     0.43s  3.67% 39.93%      2.65s 22.61%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb/memdb.(*DB).findGE
     0.37s  3.16% 43.09%      2.38s 20.31%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb.(*iComparer).Compare
     0.30s  2.56% 45.65%      0.30s  2.56%  runtime.memmove
     0.24s  2.05% 47.70%      1.54s 13.14%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb.(*iComparer).uCompare
     0.23s  1.96% 49.66%      0.50s  4.27%  runtime.scanobject
     0.22s  1.88% 51.54%      0.32s  2.73%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb.internalKey.ukey
     0.21s  1.79% 53.33%      0.56s  4.78%  bytes.(*Buffer).Write
     0.19s  1.62% 54.95%      1.30s 11.09%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb/comparer.(*bytesComparer).Compare
     0.18s  1.54% 56.48%      1.11s  9.47%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb/comparer.bytesComparer.Compare
     0.18s  1.54% 58.02%      0.18s  1.54%  runtime.duffcopy
     0.17s  1.45% 59.47%      0.17s  1.45%  runtime.heapBitsForObject
     0.14s  1.19% 60.67%      0.68s  5.80%  github.com/hyperledger/burrow/vendor/github.com/tendermint/go-wire.WriteTo
     0.14s  1.19% 61.86%      0.14s  1.19%  runtime.heapBitsSetType
     0.13s  1.11% 62.97%      0.15s  1.28%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb.internalKey.num
     0.13s  1.11% 64.08%      2.08s 17.75%  github.com/hyperledger/burrow/vendor/golang.org/x/crypto/ripemd160.(*digest).Write
     0.12s  1.02% 65.10%      0.12s  1.02%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb.internalKey.assert
     0.12s  1.02% 66.13%      0.64s  5.46%  github.com/hyperledger/burrow/vendor/github.com/tendermint/merkleeyes/iavl.(*IAVLNode).set
     0.11s  0.94% 67.06%      0.11s  0.94%  runtime.greyobject
     0.11s  0.94% 68.00%      0.19s  1.62%  runtime.mapassign_faststr
     0.11s  0.94% 68.94%      0.41s  3.50%  sort.doPivot
     0.10s  0.85% 69.80%      0.89s  7.59%  runtime.newobject
     0.09s  0.77% 70.56%      0.67s  5.72%  github.com/hyperledger/burrow/vendor/github.com/tendermint/go-wire.WriteVarint
     0.09s  0.77% 71.33%      1.63s 13.91%  github.com/hyperledger/burrow/vendor/golang.org/x/crypto/ripemd160.(*digest).Sum
     0.09s  0.77% 72.10%      0.11s  0.94%  runtime.mapdelete_faststr
     0.09s  0.77% 72.87%      0.09s  0.77%  runtime.memclrNoHeapPointers
     0.08s  0.68% 73.55%      0.51s  4.35%  github.com/hyperledger/burrow/vendor/github.com/tendermint/go-wire.WriteByte
     0.08s  0.68% 74.23%      4.67s 39.85%  github.com/hyperledger/burrow/vendor/github.com/tendermint/merkleeyes/iavl.(*IAVLNode).save
     0.08s  0.68% 74.91%      0.08s  0.68%  runtime.aeshashbody
     0.08s  0.68% 75.60%      0.18s  1.54%  runtime.deferreturn
     0.07s   0.6% 76.19%      0.25s  2.13%  bytes.(*Buffer).grow
     0.07s   0.6% 76.79%      9.50s 81.06%  github.com/hyperledger/burrow/execution.(*BlockCache).Sync
     0.07s   0.6% 77.39%      0.07s   0.6%  runtime.freedefer
     0.07s   0.6% 77.99%      0.07s   0.6%  runtime.newdefer
     0.07s   0.6% 78.58%      0.07s   0.6%  sync.(*Mutex).Lock
     0.06s  0.51% 79.10%      0.34s  2.90%  github.com/hyperledger/burrow/binary.(*Tuple256Slice).Less
     0.06s  0.51% 79.61%      0.21s  1.79%  github.com/hyperledger/burrow/vendor/github.com/syndtr/goleveldb/leveldb/table.(*blockWriter).append
```

## Diagnosis
The one to watch is `BlockCache.Sync()` my least favourite piece of code that still remains in the refactored Burrow. It's a ghastly bit of code and it looks like it is repeatedly hitting expensive merkle tree and database calls. I need a bit more time to look into why but this cache was scheduled for annihilation very soon I just had been planning to do some other integration work soon.

BlockCache holds a cumulative cache of account state between blocks preserved to speed up EVM operations. It is not meant to mark cache lines as dirty if they have not been modified, but it looks like it is doing some hashing and database ops for every account it has seen every block which is why you see a cumulative slowdown as it repeat unnecessary work.

## Temporary proof-of-concept fix
I have what I think is a proof-of-concept fix here (**please try this version of Burrow and tell me if the issue seems to be resolved**):

https://github.com/hyperledger/burrow/tree/perfo

(it is branched from `2748007` of Burrow which your reproduction is based - this is because there are some minor breakages between the released version of legacy-contracts and latest develop).

See the commit implementing the 'fix' here: https://github.com/hyperledger/burrow/commit/a0f877d707f5110bfa16fe75bdeeb4a6f1762957

For me (with one quick test) this seems to return us to a steady state of ~1% CPU usage when the chain is just making empty blocks as we would expect.

## Real fix
The real fix is to rewrite `BlockCache` which already does other badness by writing directly to database rather than through the proper interfaces. This is on my agenda within the next few weeks, but I also need to introduces some features so we may need to adopt the temporary fix in the interim. We will also be updating our merkle tree implementation for other stability and speed improvements.

**Comment by Gaurav094** [comment]
Hey Team,
When I'm taking pull of perfo branch and running cmd "make build"
I get few error. PFA...
![makebuildofdevelop](https://user-images.githubusercontent.com/36070107/37812778-cc1bf942-2e87-11e8-9ab0-89afd591868a.png)

**Comment by zramsay** [comment]
try `make ensure_vendor` beforehand

**Comment by Gaurav094** [comment]
Thanks, that worked for me.
But till now I was using master branch  and now after switching to perfo branch I was able to follow the instruction given in README. But how to get the private key of created account. Because in previous workflow of mine I was using monax through which address.json file was being created where I use to find the Private key to create Manager object using @monax/legacy-contract.js to deploy the contract?

**Comment by compleatang** [comment]
@pascalverlinden when you are able to test develop branch can you confirm this issue is fixed for y'all. We isolated and added the permanent fix so I'd like to close the ticket but will wait on y'all to confirm when you are able.

**Comment by silasdavis** [comment]
I'm going to close this ticket now as I think it is fixed. If not we can just re-open

**Comment by slavitch** [comment]
I'm seeing a 100% CPU issue.  Burrow 0.29.5 pegs 100% CPU when running a search through smart contracts, as follows

  function list_TA_Email(string memory _TA_Email) view public returns (string memory List) {
    string memory d = '{';
    uint c = 0;
    while (contracts[c].init){
      if (JsmnSolLib.strCompare(_TA_Email,contracts[c].TA_Email) == 0){
        ContractTemplate storage con = contracts[c];
        string memory i = '';
        i = JSONish_string_append(i,'"',JsmnSolLib.uint2str(c),'":');

        d = JSONish_string_append(d,i,con.JSON_RAW,'');
        c+=1;
        if (contracts[c].init){
          d = JSONish_string_append(d,',','','');
        }
      }
    }
    d = JSONish_string_append(d,'}','','');
    return(d);
  }

**Comment by silasdavis** [comment]
@slavitch firstly thanks for providing this report.

Would you mind opening this as a separate issue - this is an ancient issue and I don't want to lump all performance issues into a single old ticket?

Can you provide a bit of context. Looks like you have solidity code there, can you answer:

- How are you deploying the contract
- How are you performing the call (e.g. burrow deploy/js/GRPC directly)
- What is the steady state - e.g. the tx eventually is confirmed and processed, Burrow crashes, 

Can you also provide the `JsmnSolLib`
